inc/saiport.h

```diff
@@ -171,11 +171,17 @@ typedef enum _sai_port_fec_mode_t
     /** No FEC */
     SAI_PORT_FEC_MODE_NONE,
 
-    /** Enable RS-FEC - 25G, 50G, 100G ports */
+    /** Enable RS-528 FEC - 25G, 50G, 100G ports */
     SAI_PORT_FEC_MODE_RS,
```

Comment:
Instead of overloading SAI_PORT_FEC_MODE_RS with RS-528 FEC, can we add SAI_PORT_FEC_MODE_RS528 instead so that the backward compatibility is not broken?

Reply:
To avoid ambiguity in the readback (if application sets SAI_PORT_FEC_MODE_RS and SAI adapter actually configures RS-544, what is the correct readback?), I think we either need to redefine this or define a new port attribute that explicitly defines the FEC mode.

Since breaking backward compatibility is a concern here, I'm leaning toward the latter based on the conversation today - creating SAI_PORT_ATTR_DETAILED_FEC_MODE which allows for detailed control while leaving the existing enum alone. The problems I see with doing this are:
1) we need a way to enforce that only one or the other is set (or specify which is used)
2) we probably need to create corresponding attributes for all uses of sai_port_fec_mode_t (SAI_PORT_ATTR_SUPPORTED_FEC_MODE, SAI_PORT_ATTR_REMOTE_ADVERTISED_FEC_MODE, SAI_PORT_ATTR_ADVERTISED_FEC_MODE, SAI_PORT_ATTR_FEC_MODE)

Thoughts on this? @lguohan @JaiOCP @SathishKumarTR-Dell @dgsudharsan @rlhui

Reply:
SAI_PORT_FEC_MODE_RS has been using for RS528, RS544, RS272 and RS544_2XN.

Reply:
Yes. The problem I'm trying to solve is that the mapping of SAI_PORT_FEC_MODE_RS is to the actual FEC type is ambiguous in at least one case (200G PAM4 can use either RS544 or RS544_2XN), and there are cases when using optics with built-in gearbox that we may need to use different FEC configuration (I believe 100G/lane DR8 communicating with 50G/lane DR4 will be an example of this).

Reply:
The comment should be changed to the original one (i.e. remove 528 from the comment)

Reply:
Are you suggesting that we have all of the RS-FEC types in addition to the legacy "FEC_MODE_RS"?
SAI_PORT_FEC_MODE_RS
SAI_PORT_FEC_MODE_RS528
SAI_PORT_FEC_MODE_RS544
SAI_PORT_FEC_MODE_RS544_2X_INTERLEAVED

I think I'd be OK with that, but we would need to document that if FEC_MODE_RS is set then the actually selected FEC mode (one of RS528,544,etc) would be returned? This is similar to what's being proposed by @jleveque in https://github.com/opencomputeproject/SAI/pull/1233.

Reply:
Yes.  This sounds OK.  Thanks.

Reply:
Please make sure to remove change in line 174.

Reply:
Are there any concerns around warmboot reconciliation or SYNCD_INSPECT_ASIC at the SONiC (or other application) level if we start introducing attributes that read back differently from what is set? It appears to me that this type of setting (write one value, read back a different one) would require special handling in those cases.

Reply:
Based on the conversations above (needing to preserve the existing behavior) and my concerns about reading back a different value than was programmed, I made the change to a new enum for detailed FEC control. Please take a look and let me know if this is more acceptable.

Reply:
I believe that this is resolved at this point.